leshan-server-core/src/main/java/org/eclipse/leshan/server/queue/PresenceServiceImpl.java

```diff
@@ -0,0 +1,186 @@
+ * Copyright (c) 2017 Bosch Software Innovations GmbH and others.
+ *
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * and Eclipse Distribution License v1.0 which accompany this distribution.
+ *
+ * The Eclipse Public License is available at
+ *    http://www.eclipse.org/legal/epl-v10.html
+ * and the Eclipse Distribution License is available at
+ *    http://www.eclipse.org/org/documents/edl-v10.html.
+ *
+ * Contributors:
+ *     Bosch Software Innovations GmbH - initial API
+ *     RISE SICS AB - added more features 
+ *******************************************************************************/
+package org.eclipse.leshan.server.queue;
+
+import java.util.List;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.CopyOnWriteArrayList;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+
+import org.eclipse.leshan.server.registration.Registration;
+
+/**
+ * Tracks the status of each LWM2M client registered with Queue mode binding. Also ensures that the
+ * {@link PresenceListener} are notified on state changes only for those LWM2M clients registered using Queue mode
+ * binding.
+ * 
+ * @see Presence
+ */
+public final class PresenceServiceImpl implements PresenceService {
+
+    private final ConcurrentMap<String, PresenceStatus> clientStatusList = new ConcurrentHashMap<>();
+    private final List<PresenceListener> listeners = new CopyOnWriteArrayList<>();
+    ScheduledExecutorService clientTimersExecutor = Executors.newSingleThreadScheduledExecutor();
+
+    @Override
+    public void addListener(PresenceListener listener) {
+        listeners.add(listener);
+    }
+
+    @Override
+    public void removeListener(PresenceListener listener) {
+        listeners.remove(listener);
+    }
+
+    @Override
+    public boolean isClientAwake(Registration registration) {
+        if (!clientStatusList.containsKey(registration.getEndpoint())) {
+            return false;
+        }
+
+        PresenceStatus presenceStatus = clientStatusList.get(registration.getEndpoint());
+        return presenceStatus.isClientAwake();
+
+    }
+
+    /**
+     * Set the state of the client identified by registration as {@link Presence#AWAKE}
+     * 
+     * @param reg the client's registration object
+     */
+    public void setAwake(Registration reg) {
+        if (reg.usesQueueMode()) {
+            if (!clientStatusList.containsKey(reg.getEndpoint())) {
+                createPresenceStatusObject(reg);
+            }
+
+            boolean stateChanged = false;
+            synchronized (this) {
+                stateChanged = clientStatusList.get(reg.getEndpoint()).setAwake();
+                startClientAwakeTimer(reg);

```

Comment:
needed in the synchronized block?

Reply:
I would like to post pone polish about threadsafeness for the next version but I will do a first check now.